### PR TITLE
Feat/gho steward

### PR DESCRIPTION
**Pull request by foodaka**

### CODE REVIEW

**Review by ghost — APPROVED** [review]
LGTM

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol [hunk 1]
```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+import {AaveV3Ethereum, IACLManager} from 'aave-address-book/AaveV3Ethereum.sol';
+import {IProposalGenericExecutor} from 'aave-helpers/interfaces/IProposalGenericExecutor.sol';
+import {AaveGovernanceV2} from 'aave-address-book/AaveGovernanceV2.sol';
+import {GhoSteward} from 'gho-core/contracts/misc/GhoSteward.sol';
+import {IGhoToken} from 'gho-core/contracts/gho/interfaces/IGhoToken.sol';
+
+library Create2Helpers {
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: I do not think the library is needed. My personal preference would be moving these functions to the payload as `internal`.

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol [hunk 2]
```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+import {AaveV3Ethereum, IACLManager} from 'aave-address-book/AaveV3Ethereum.sol';
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: missing a newline here

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+ * - Grants role of BucketManager to GhoSteward
+ * Relevant governance links:
+ * - Governance: https://governance.aave.com/t/arfc-gho-steward-agile-parameter-changes/13922
+ * - Snapshot: https://snapshot.org/#/aave.eth/proposal/0x67fa551993a94b801018d02154f3d4f27e29bea51fe6e862686de6bf9ee650af
+ */
+
+contract AaveV3ListGhoStewardsPayload is IProposalGenericExecutor {
+  event GhoStewardDeployed(address ghoSteward);
+
+  bytes public GHO_STEWARD_BYTECODE;
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: This cannot be a variable in the storage. Make it `immutable`.

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+ * @author @aave - Aave Co
+ * @dev This proposal deploys GhoSteward, which consists of a set of actions.
+ * - Deployment of GhoSteward and transfers ownership to Aave DAO (short executor)
+ * - Grants role of PoolAdmin to GhoSteward
+ * - Grants role of BucketManager to GhoSteward
+ * Relevant governance links:
+ * - Governance: https://governance.aave.com/t/arfc-gho-steward-agile-parameter-changes/13922
+ * - Snapshot: https://snapshot.org/#/aave.eth/proposal/0x67fa551993a94b801018d02154f3d4f27e29bea51fe6e862686de6bf9ee650af
+ */
+
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: Remove this line

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        bytecode,
+        abi.encode(poolAddressProvider, ghoToken, riskCouncil, shortExecutor)
+      );
+  }
+}
+
+/**
+ * @title Gho Stewards AIP
+ * @author @aave - Aave Co
+ * @dev This proposal deploys GhoSteward, which consists of a set of actions.
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]:
```suggestion
 * @author AaveCompanies
 * @dev This proposal enables GhoSteward for agile parameter changes in GHO. It consists of a set of actions:
```

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+ */
+
+contract AaveV3ListGhoStewardsPayload is IProposalGenericExecutor {
+  event GhoStewardDeployed(address ghoSteward);
+
+  bytes public GHO_STEWARD_BYTECODE;
+  address public immutable GHO_STEWARD;
+
+  constructor() {
+    GHO_STEWARD_BYTECODE = Create2Helpers.getBytecode(
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: This intermediate variable is not needed imo.

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+      address(AaveV3Ethereum.POOL_ADDRESSES_PROVIDER),
+      GHO_TOKEN,
+      RISK_COUNCIL,
+      AaveGovernanceV2.SHORT_EXECUTOR
+    );
+
+    // ------------------------------------------------
+    // 2. Set roles of PoolAdmin and BucketManager to GhoSteward
+    // ------------------------------------------------
+    AaveV3Ethereum.ACL_MANAGER.addPoolAdmin(address(deployedGhoSteward));
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: unneeded casting here

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+    IGhoToken iGhoToken = IGhoToken(GHO_TOKEN);
+    iGhoToken.grantRole(iGhoToken.BUCKET_MANAGER_ROLE(), address(deployedGhoSteward));
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: unneeded casting here

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    // ------------------------------------------------
+    // 1. Deployment of GhoSteward
+    // ------------------------------------------------
+    address deployedGhoSteward = Create2Helpers.deployGhoSteward(
+      address(AaveV3Ethereum.POOL_ADDRESSES_PROVIDER),
+      GHO_TOKEN,
+      RISK_COUNCIL,
+      AaveGovernanceV2.SHORT_EXECUTOR
+    );
+
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: Add a require here, to confirm deployedGhoSteward == precomputed address please

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.t.sol [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+      RISK_COUNCIL,
+      AaveGovernanceV2.SHORT_EXECUTOR
+    );
+
+    GHO_STEWARD = Create2Helpers.getAddress(GHO_STEWARD_BYTECODE);
+  }
+
+  function testGhoStewardRoles() public {
+    bool beforeGhoStewardAdmin = AaveV3Ethereum.ACL_MANAGER.isPoolAdmin(address(GHO_STEWARD));
+    assertEq(false, beforeGhoStewardAdmin);
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.t.sol]: Change to `assertTrue`

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.t.sol [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+    bool afterHasBucketManagerRole = iGhoToken.hasRole(
+      iGhoToken.BUCKET_MANAGER_ROLE(),
+      address(GHO_STEWARD)
+    );
+
+    assertEq(true, afterHasBucketManagerRole);
+  }
+
+  function testUpdateBorrowRates() public {
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.t.sol]: would be nice to add a fail scenario, where the signer is not the riskCouncil (same for `updateBucketCapacity`)

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+ * @dev This proposal deploys GhoSteward, which consists of a set of actions.
+ * - Deployment of GhoSteward and transfers ownership to Aave DAO (short executor)
+ * - Grants role of PoolAdmin to GhoSteward
+ * - Grants role of BucketManager to GhoSteward
+ * Relevant governance links:
+ * - Governance: https://governance.aave.com/t/arfc-gho-steward-agile-parameter-changes/13922
+ * - Snapshot: https://snapshot.org/#/aave.eth/proposal/0x67fa551993a94b801018d02154f3d4f27e29bea51fe6e862686de6bf9ee650af
+ */
+
+contract AaveV3ListGhoStewardsPayload is IProposalGenericExecutor {
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: Let's change the name to `GhoAaveV3GhoSteward`.
All proposals about Gho can start with the preffix `Gho`

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol (hunk 2 above)
Comment (foodaka) [reply src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: updated 6990be2287dadbeeda3e2519970957915a16896f

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol (hunk 9 above)
Comment (foodaka) [reply src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: added 6990be2287dadbeeda3e2519970957915a16896f

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.t.sol (hunk 11 above)
Comment (foodaka) [reply src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.t.sol]: added two new tests 6990be2287dadbeeda3e2519970957915a16896f

File: src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol [hunk 13]
```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+
+import {AaveV3Ethereum, IACLManager} from 'aave-address-book/AaveV3Ethereum.sol';
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/AaveV3ListGhoStewards_20230507_Payload.sol]: `IACLManager` import not needed

File: src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol [hunk 14]
```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.16;
+import 'forge-std/Test.sol';
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol]: newline here

File: src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+import {IGhoSteward} from 'gho-core/contracts/misc/interfaces/IGhoSteward.sol';
+import {IPoolDataProvider} from 'aave-v3-core/contracts/interfaces/IPoolDataProvider.sol';
+
+import {IGhoDiscountRateStrategy} from 'gho-core/contracts/facilitators/aave/interestStrategy/interfaces/IGhoDiscountRateStrategy.sol';
+import {GhoInterestRateStrategy} from 'gho-core/contracts/facilitators/aave/interestStrategy/GhoInterestRateStrategy.sol';
+import {IGhoToken} from 'gho-core/contracts/gho/interfaces/IGhoToken.sol';
+
+contract GhoAaveV3GhoSteward_20230507_Payload_Test is ProtocolV3TestBase {
+  GhoAaveV3GhoSteward public proposalPayload;
+  address immutable ZERO_ADDRESS = 0x0000000000000000000000000000000000000000;
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol]: you can remove and use `address(0)` where needed instead

File: src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  function testUpdateBucketCapacity() public {
+    vm.startPrank(proposalPayload.RISK_COUNCIL());
+
+    IPoolDataProvider iPoolDataProvider = IPoolDataProvider(
+      AaveV3Ethereum.AAVE_PROTOCOL_DATA_PROVIDER
+    );
+
+    (address GHO_ATOKEN, , ) = iPoolDataProvider.getReserveTokensAddresses(
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol]: use lower case for `ghoAToken`

File: src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.sol [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+ * @author @aave - AaveCompanies
+ * @dev This proposal enables GhoSteward for agile parameter changes in GHO. It consists of a set of actions:
+ * - Deployment of GhoSteward and transfers ownership to Aave DAO (short executor)
+ * - Grants role of PoolAdmin to GhoSteward
+ * - Grants role of BucketManager to GhoSteward
+ * Relevant governance links:
+ * - Governance: https://governance.aave.com/t/arfc-gho-steward-agile-parameter-changes/13922
+ * - Snapshot: https://snapshot.org/#/aave.eth/proposal/0x67fa551993a94b801018d02154f3d4f27e29bea51fe6e862686de6bf9ee650af
+ */
+contract GhoAaveV3GhoSteward is IProposalGenericExecutor {
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.sol]: You may need to add date here too, same name as the filename

File: src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol [hunk 18]
```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.16;
+import 'forge-std/Test.sol';
+import {AaveV3Ethereum, AaveV3EthereumAssets} from 'aave-address-book/AaveV3Ethereum.sol';
+import {ProtocolV3TestBase, InterestStrategyValues, ReserveConfig} from 'aave-helpers/ProtocolV3TestBase.sol';
```
Comment (miguelmtzinf) [inline src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol]: review imports, I think some of them are unnecessary

File: src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol (hunk 16 above)
Comment (foodaka) [reply src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol]: because its not a constant set in the contract?

Comment (foodaka) [reply src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol]: resolved 3f169bdfd649faeca2ddbfbe92d00350ee5be7c5

File: src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol (hunk 15 above)
Comment (foodaka) [reply src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol]: resolved 3f169bdfd649faeca2ddbfbe92d00350ee5be7c5

File: src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.sol (hunk 17 above)
Comment (foodaka) [reply src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.sol]: updated to in `GhoAaveV3GhoSteward_20230507_Payload` 3f169bdfd649faeca2ddbfbe92d00350ee5be7c5

File: src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol (hunk 18 above)
Comment (foodaka) [reply src/AaveV3ListGhoStewards_20230507/GhoAaveV3GhoSteward_20230507_Payload.t.sol]: resolved in 3f169bdfd649faeca2ddbfbe92d00350ee5be7c5

**Review by miguelmtzinf — APPROVED** [review]

**Pull request merged**